opbeat.go

```diff
@@ -0,0 +1,443 @@
+}
+
+// Configure Opbeat application credentials. These can be found when viewing
+// your application settings on the Opbeat website.
+func (opbeat *Opbeat) Credentials(organizationId, appId, secretToken string) {
+	opbeat.organizationId = organizationId
+	opbeat.appId = appId
+	opbeat.secretToken = secretToken
+}
+
+// HTTP middleware handler that automatically will log any paniced error to
+// Opbeat.
+func (opbeat *Opbeat) Handler(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		defer func() {
+			if err := recover(); err != nil {
+				if err, ok := err.(error); ok {
+					opbeat.CaptureErrorWithRequest(err, r)
+				}
+			}
+		}()
+		h.ServeHTTP(w, r)
+	})
+}
+
+// Captures an error and sends the log to Opbeat.
+func (opbeat *Opbeat) CaptureError(err error) error {
+	return opbeat.CaptureErrorSkip(err, 1, nil)
+}
+
+// Captures an error along with a `*http.Request` and sends the log to Opbeat
+// enriched with information specific to that request. Useful when using Opbeat
+// in a http application.
+func (opbeat *Opbeat) CaptureErrorWithRequest(e error, r *http.Request) error {
+	headers := make(map[string]string)
+	for k, v := range r.Header {
+		headers[k] = v[len(v)-1]
+	}
+
+	scheme := ""
+	if r.TLS != nil {
+		scheme = "https"
+	} else {
+		scheme = "http"
+	}
+
+	http := map[string]interface{}{
+		"url":     scheme + "://" + r.Host + r.URL.String(),
+		"method":  r.Method,
+		"headers": headers,
+	}
+
+	return opbeat.CaptureErrorSkip(e, 4, map[string]interface{}{
+		"http": http,
+	})
+}
+
+// Captures an error and skips `skip` amount of frames in the stacktrace, this
+// is useful to stop logging library type frames to Opbeat. Also takes a map of
+// other interfaces that are written to Opbeat as a part of the log. Please take
+// care that any values in this map can be marshalled into JSON.
+func (opbeat *Opbeat) CaptureErrorSkip(e error, skip int, options map[string]interface{}) error {
+	stacktrace, err := stacko.NewStacktrace(3 + skip)
+	if err != nil {
+		return err
+	}
+
+	p, err := newPacket(e.Error(), stacktrace)
+	if err != nil {
+		return err
+	}
+
+	p.Revision = opbeat.Revision
+	p.Level = Error
+	p.Logger = opbeat.LoggerName
+
+	if http, ok := options["http"].(map[string]interface{}); ok {
+		p.HTTP = http
+	}
+
+	opbeat.queue(p)
+
+	return nil
+}
+
+// Captures a message along with a level indicating the severity of the message.
+func (opbeat *Opbeat) CaptureMessage(message string, l Level) error {
+	p, err := newPacket(message, nil)
+	if err != nil {
+		return err
+	}
+
+	p.Level = l
+
+	opbeat.queue(p)
+
+	return nil
+}
+
+// Waits for all packets to send.
+func (opbeat *Opbeat) Wait() {
+	opbeat.wait.Wait()
+}
+
+// Starts a goroutine which listens on the main channel and sends any packets it
+// receives to Opbeat. The goroutine will exit if the client's `.Close` function
+// is called.
+func (opbeat *Opbeat) Start() {
+	opbeat.packets = make(chan *packet)
```

Comment:
This channel should probably be buffered so callers of `CaptureX` won't block if another request to Opbeat is in progress.